modular/gater/object_handler.go

```diff
@@ -985,6 +990,494 @@ func (g *GateModular) getObjectByUniversalEndpointHandler(w http.ResponseWriter,
+		return
+	}
+
+	if err = g.checkSPAndBucketStatus(reqCtx.Context(), reqCtx.bucketName, reqCtx.account); err != nil {
+		log.CtxErrorw(reqCtx.Context(), "resumable put object failed to check sp and bucket status", "error", err)
+		return
+	}
+	if isUpdate {
+		authenticated, err = g.baseApp.GfSpClient().VerifyAuthentication(reqCtx.Context(), coremodule.AuthOpTypeAgentUpdateObject,
+			reqCtx.Account(), reqCtx.bucketName, reqCtx.objectName)
+	} else {
+		authenticated, err = g.baseApp.GfSpClient().VerifyAuthentication(reqCtx.Context(), coremodule.AuthOpTypeAgentPutObject,
+			reqCtx.Account(), reqCtx.bucketName, reqCtx.objectName)
+	}
+	if !authenticated {
+		log.CtxErrorw(reqCtx.Context(), "no permission to operate")
+		err = ErrNoPermission
+		return
+	}
+
+	approvalMsg, err = hex.DecodeString(r.Header.Get(GnfdUnsignedApprovalMsgHeader))
+	if err != nil {
+		log.Errorw("failed to parse approval header",
+			"approval", r.Header.Get(GnfdUnsignedApprovalMsgHeader))
+		err = ErrDecodeMsg
+		return
+	}
+	fingerprint = commonhash.GenerateChecksum(approvalMsg)
+	startTime := time.Now()
+
+	if isUpdate {
+		objectInfo, err = g.baseApp.Consensus().QueryObjectInfo(reqCtx.ctx, reqCtx.bucketName, reqCtx.objectName)
+		if err != nil {
+			log.CtxErrorw(reqCtx.Context(), "failed to get object info from consensus", "error", err)
+			err = ErrConsensusWithDetail("failed to get object info from consensus, error: " + err.Error())
+			return
+		}
+
+		if !objectInfo.IsUpdating {
+			msg := &storagetypes.MsgDelegateUpdateObjectContent{
+				Updater:     reqCtx.account,
+				BucketName:  reqCtx.bucketName,
+				ObjectName:  reqCtx.objectName,
+				PayloadSize: payloadSize,
+				ContentType: contentType,
+			}
+			txHash, err = g.baseApp.GfSpClient().DelegateUpdateObjectContent(reqCtx.ctx, msg)
+			if err != nil {
+				log.CtxErrorw(reqCtx.ctx, "failed to delegate update object", "error", err)
+				return
+			}
+		}
+
+	} else {
+		var visibilityInt int64
+		visibilityStr := queryParams.Get("visibility")
+		visibilityInt, err = strconv.ParseInt(visibilityStr, 10, 32)
+		if err != nil {
+			return
+		}
+		visibility = storagetypes.VisibilityType(visibilityInt)
+		if visibility == storagetypes.VISIBILITY_TYPE_UNSPECIFIED {
+			visibility = storagetypes.VISIBILITY_TYPE_INHERIT // set default visibility type
+		}
+		task := &gfsptask.GfSpDelegateCreateObjectApprovalTask{}
+		task.InitApprovalDelegateCreateObjectTask(reqCtx.Account(), &storagetypes.MsgDelegateCreateObject{
+			Operator:       g.baseApp.OperatorAddress(),
+			Creator:        reqCtx.account,
+			BucketName:     reqCtx.bucketName,
+			ObjectName:     reqCtx.objectName,
+			PayloadSize:    payloadSize,
+			ContentType:    contentType,
+			Visibility:     visibility,
+			RedundancyType: storagetypes.REDUNDANCY_EC_TYPE,
+		}, fingerprint, g.baseApp.TaskPriority(task))
+
+		objectInfo, err = g.baseApp.Consensus().QueryObjectInfo(reqCtx.ctx, reqCtx.bucketName, reqCtx.objectName)
+		if objectInfo == nil {
```

Comment:
can not assert like this. if there is RPC error, this would still send the tx to chain. So you need to check if its exactly "no such object" error from chain.